File: Java/problem_0035_search_insert_position.java
package Java;

import org.junit.Test;
import static org.junit.Assert.assertEquals;

public class problem_0035_search_insert_position {
    public int searchInsert(int[] nums, int target) {
        int leftPointer = 0;
        int rightPointer = nums.length - 1;
        while (leftPointer <= rightPointer) {
            int mid = leftPointer + (rightPointer - leftPointer) / 2;
            if (nums[mid] == target) {
                return mid;
            } else if (nums[mid] < target) {
                leftPointer = mid + 1;
            } else {
                rightPointer = mid - 1;
            }
        }
        return leftPointer;
    }

    @Test
    public void testSearchInsert() {
        int[] testCase1 = { 1, 3, 5, 6 };
        assertEquals(2, searchInsert(testCase1, 5));
        assertEquals(1, searchInsert(testCase1, 2));
        assertEquals(4, searchInsert(testCase1, 7));
    }
}

/*
 * Explanation
 * Similar to Binary Search
 * 
 * Time: O(logn)
 * Space: O(1)
 */

File: csharp/solution_and_tests/binary_search/problem_0035_search_insert_position.cs
/*
 * https://leetcode.com/problems/search-insert-position/
 * Explanation:
 * 
 * Similar to Binary Search https://leetcode.com/problems/binary-search/
 * 
 * Time: O(logn)
 * Space: O(1)
 */

using NUnit.Framework;

namespace Search_Insert_Position
{
    public class Solution
    {
        public static int SearchInsert(int[] nums, int target)
        {
            int leftPointer = 0;
            int rightPointer = nums.Length - 1;
            while (leftPointer <= rightPointer)
            {
                int mid = leftPointer + (rightPointer - leftPointer) / 2;
                if (nums[mid] == target)
                {
                    return mid;
                }
                else if (nums[mid] < target)
                {
                    leftPointer = mid + 1;
                }
                else
                {
                    rightPointer = mid - 1;
                }
            }
            return leftPointer;
        }
    }
    public class Tests
    {
        [Test]
        [TestCase(new int[] { 1, 3, 5, 6 }, 5, ExpectedResult = 2)]
        [TestCase(new int[] { 1, 3, 5, 6 }, 2, ExpectedResult = 1)]
        [TestCase(new int[] { 1, 3, 5, 6 }, 7, ExpectedResult = 4)]
        public int TestSearchInsert(int[] nums, int target)
        {
            return Solution.SearchInsert(nums, target);
        }
    }
}